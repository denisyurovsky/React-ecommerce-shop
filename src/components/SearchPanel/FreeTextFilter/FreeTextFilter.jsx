import SearchIcon from '@mui/icons-material/Search';
import { OutlinedInput, Button, Box } from '@mui/material';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import styles from './FreeTextFilter.module.scss';

const FreeTextFilter = ({ className, setFilterProperties }) => {
  const [value, setValue] = useState('');
  const handleOnSubmit = (event) => {
    event.preventDefault();
    const formData = new FormData(event.target);
    const searchValue = formData.get('searchValue');

    setFilterProperties((prev) => ({
      ...prev,
      searchValue: searchValue,
      currentPage: 1,
    }));
  };

  return (
    <Box
      component="form"
      onSubmit={handleOnSubmit}
      className={classNames({ [styles.wrapper]: true, [className]: true })}
    >
      <OutlinedInput
        value={value}
        name="searchValue"
        sx={{
          borderTopRightRadius: '0px',
          borderBottomRightRadius: '0px',
          flexGrow: 1,
        }}
        onChange={(e) => setValue(e.target.value)}
        className={styles.input}
        placeholder="Search by advertisement"
      />
      <Button
        type="submit"
        sx={{
          borderTopLeftRadius: '0px',
          borderBottomLeftRadius: '0px',
        }}
        variant="contained"
        color="primary"
      >
        <SearchIcon color="white" />
      </Button>
    </Box>
  );
};

FreeTextFilter.propTypes = {
  setFilterProperties: PropTypes.func.isRequired,
  className: PropTypes.string,
};

export default FreeTextFilter;
