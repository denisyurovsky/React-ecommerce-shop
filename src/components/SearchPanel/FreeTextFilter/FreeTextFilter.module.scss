.wrapper {
  display: flex;
}

.icon {
  color: white;
}
