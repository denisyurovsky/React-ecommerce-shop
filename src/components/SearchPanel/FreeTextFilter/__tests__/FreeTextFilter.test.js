import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import FreeTextFilter from '../FreeTextFilter';

describe('Search component', () => {
  describe('snapshots', () => {
    it('renders a valid snapshot', () => {
      const setFilterProperties = jest.fn();
      const { asFragment } = render(
        <FreeTextFilter setFilterProperties={setFilterProperties} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('main group of tests', () => {
    it('typing in search works properly', () => {
      const setFilterProperties = jest.fn();

      render(<FreeTextFilter setFilterProperties={setFilterProperties} />);
      const inputElement = screen.getByPlaceholderText(
        'Search by advertisement'
      );

      expect(inputElement).toBeInTheDocument();
      userEvent.type(inputElement, '');
      expect(inputElement).toHaveValue('');
      userEvent.type(inputElement, 'test');
      expect(inputElement).toHaveValue('test');
      userEvent.type(
        inputElement,
        '{backspace}{backspace}{backspace}{backspace}'
      );
      expect(inputElement).toHaveValue('');
    });
    it('search button works properly', () => {
      let data;
      const setFilterProperties = jest.fn((func) => (data = func()));

      render(<FreeTextFilter setFilterProperties={setFilterProperties} />);
      const inputElement = screen.getByPlaceholderText(
        'Search by advertisement'
      );
      const button = screen.getByRole('button');

      userEvent.type(inputElement, 'test');
      userEvent.click(button);
      expect(data.searchValue).toEqual('test');
    });
  });
});
