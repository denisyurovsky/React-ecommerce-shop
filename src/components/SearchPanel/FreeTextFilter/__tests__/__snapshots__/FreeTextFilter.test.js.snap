// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component snapshots renders a valid snapshot 1`] = `
<DocumentFragment>
  <form
    class="wrapper undefined MuiBox-root css-0"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary input css-ocdt0h-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        name="searchValue"
        placeholder="Search by advertisement"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-nnbavb"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ik5mn7-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorWhite MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
